infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/DockerInternalRuntime.java

```diff
@@ -240,120 +231,46 @@ private void startMachine(String name,
-    private void checkServerReadiness(String machineName,
-                                      String serverRef,
-                                      String serverUrl)
-            throws InfrastructureException {
 
-        if (!livenessChecksPaths.containsKey(serverRef)) {
-            return;
-        }
-        String livenessCheckPath = livenessChecksPaths.get(serverRef);
-        HttpConnectionChecker checker = "terminal".equals(serverRef) ? new TerminalHttpConnectionChecker()
-                                                                     : new HttpConnectionChecker();
-        URL url;
-        try {
-            url = UriBuilder.fromUri(serverUrl)
-                            .replacePath(livenessCheckPath)
-                            .build()
-                            .toURL();
-        } catch (MalformedURLException e) {
-            throw new InternalInfrastructureException("Server " + serverRef +
-                                                      " URL is invalid. Error: " + e.getLocalizedMessage(), e);
-        }
-        // max start time 180 seconds
-        long readinessDeadLine = System.currentTimeMillis() + 3000 * 60;
-        while (System.currentTimeMillis() < readinessDeadLine) {
-            LOG.info("Checking server {} of machine {} at {}", serverRef, machineName,
-                     System.currentTimeMillis());
-            checkStartInterruption();
-            if (isHttpConnectionSucceed(url, checker)) {
-                // TODO protect with lock, from null, from exceptions
-                DockerMachine machine = startSynchronizer.getMachines().get(machineName);
-                machine.setServerStatus(serverRef, ServerStatus.RUNNING);
-
-                eventService.publish(DtoFactory.newDto(ServerStatusEvent.class)
-                                               .withIdentity(DtoConverter.asDto(identity))
-                                               .withMachineName(machineName)
-                                               .withServerName(serverRef)
-                                               .withStatus(ServerStatus.RUNNING)
-                                               .withServerUrl(serverUrl));
-                LOG.info("Server {} of machine {} started", serverRef, machineName);
-                return;
-            }
+        bootstrapperFactory.create(name, identity, dockerMachine, machineConfig.getAgents())
+                           .bootstrap();
 
-            try {
-                Thread.sleep(3000);
-            } catch (InterruptedException e) {
-                throw new InternalInfrastructureException("Interrupted");
-            }
-        }
-
-        throw new InfrastructureException("Server " + serverRef + " in machine " + machineName + " not available.");
+        new ServersReadinessChecker().check(name, dockerMachine.getServers(), new ServerReadinessHandler(name));
```

Comment:
I like it more, at least it looks like async call + wait
You don't need to pass `new ServerReadinessHandler(name)` as it can be created inside the checker, so the code would be
```java
ServersReadinessChecker checker = new ServersReadinessChecker(name, machine.getServers());
checker.startAsync();
checker.await();
```